sw/device/lib/uart.c

```diff
@@ -42,5 +42,16 @@ void uart_send_uint(uint32_t n, int bits) {
 }
 
 int uart_rcv_char(char *c) {
-  return dif_uart_byte_receive_polled(&uart0, (uint8_t *)c);
+  int ret;
+  size_t num_bytes = 0;
+  if (dif_uart_rx_bytes_available(&uart0, &num_bytes) == false) {
+    ret = -1;
+  } else {
+    if (num_bytes == 0) {
+      ret = -1;
+    } else {
+      ret = (int)dif_uart_byte_receive_polled(&uart0, (uint8_t *)c);
```

Comment:
1. The cast to `int` isn't great. Maybe a ternary would be better here.
2. The cast to `uint8_t*` is even more dangerous than the cast to `int`.
3. You should be using `dif_uart_bytes_receive` with `bytes_requested` equal to 1, because you already checked whether there was data in the FIFO. You don't need to do a blocking/polling read.
4. Do we need a `dif_uart_byte_receive` function which reads only one byte, but doesn't poll until the FIFO actually contains info?

Reply:
Thanks for your feedback @lenary , that's much appreciated.
1. Done.
2. That's not clear to me: could you please explain why a cast from `char*` to `uint8_t*` is dangerous? How should this be done?
3. Thanks, I did not notice that we have such a function. I switched to that one.
4. I don't think such a function is required. Personally, I do not like so much the general concept of the `uart_rcv_char()` function (read just a single char if possible). Maybe the whole `uart.c/h` should be reworked at some point. Anyway, this PR just aims at getting the function to work again for the examples using it.

Reply:
1. Thanks.
2. Any cast between pointer types is dangerous (The behaviour upon accessing memory through the pointer is undefined by the C standard). This one is closer to OK because you are casting to (in effect) an (unsigned) `char*`, which in C is allowed to access any kind of memory. C's rules around this are arcane, but I don't want the interface to be updated with this PR, so please may you just add a comment that the cast is dangerous and needs revisiting?
3. Thanks
4. I agree `uart.h/c` could be reworked at some point, but for the moment I agree that we should kick the can down the road until we have some users that really need it.

Reply:
Thanks @lenary for clarifying this, I put a comment in the code.